### PR TITLE
feat: :stethoscope: allow customized health check endpoint

### PR DESCRIPTION
**Pull request by Mason-Lin**
~~Get the liveness endpoint from the environment variable `HEALTH_ENDPOINT` and verify it. If the customized endpoint is invalied, it will fallback to the DEFAULT_HEALTH_ENDPOINT.~~

UPDATE 2022-10-21
----------------------------

`pypiserver` provides a default health endpoint at `/health`. It always returns
`200 Ok` if the service is up. Otherwise, it means that the service is not responsive.

In addition, `pypiserver` allows users to customize the health endpoint.

Alphanumeric characters, hyphens, forward slashes and underscores are allowed and the endpoint should not overlap with any existing routes.

Valid examples: `/healthz`, `/health/live-1`, `/api_health`, `/action/health`

### CODE REVIEW

**Comment by Mason-Lin** [comment]
Hi @dee-me-tree-or-love,
May I get your approval to run the tests workflow?

**Comment by dee-me-tree-or-love** [comment]
Hey @Mason-Lin sure! The CI's on for the PR :) Thanks for sending in the idea! I will try to leave a proper review a little later today. 👍

**Comment by Mason-Lin** [comment]
> Hey @Mason-Lin sure! The CI's on for the PR :) Thanks for sending in the idea! I will try to leave a proper review a little later today. 👍

Thank you so much!

**Comment by Mason-Lin** [comment]
Hi @dee-me-tree-or-love ,

In the beginning, I tried to copy FALLBACK_URL and add --health-endpoint to [DEFAULTS](https://github.com/pypiserver/pypiserver/blob/master/pypiserver/config.py#L78) and [RunConfig](https://github.com/pypiserver/pypiserver/blob/master/pypiserver/config.py#L660) and [run_parser](https://github.com/pypiserver/pypiserver/blob/master/pypiserver/config.py#L377)

Then I replaced [`@app.route("/health")`](https://github.com/pypiserver/pypiserver/blob/master/pypiserver/_app.py#L85) with `@app.route(config.health_endpoint)` 

I got an error while executing the test

Then I checked [`app = pypiserver.app_from_config(config)`](https://github.com/pypiserver/pypiserver/blob/master/pypiserver/__main__.py#L171)
I found that [`app_from_config`](https://github.com/pypiserver/pypiserver/blob/master/pypiserver/__init__.py#L127) is use `__import__` to `import _app` first, and then mutate global variable `_app.config = config`

Because the `bottle.route` decorator already needs `config.health_endpoint` when `import _app`, it seems that the current import order cannot get CLI arguments while creating the bottle instance.
I'm not quite sure how to solve this.
I saw a function in the bottle [`add_route`](https://github.com/pypiserver/pypiserver/blob/master/pypiserver/bottle.py#L775), it may be able to delay the time of adding endpoints but I am not very familiar with the bottle.

The above is the blockage I encountered while exploring the CLI interface.😅

I would greatly appreciate it if you kindly give me some suggestions.

**Comment by dee-me-tree-or-love** [comment]
Hey @Mason-Lin! Thanks a lot for the detailed description, this helps a lot, indeed I see the trouble 👍 I'm sorry I didn't get back to this yesterday, had some unexpected changes to my plans, but I will try to think along in the coming days and will keep you posted. 

What you describe is indeed a bit of a blocking issue, so I will try to see how to work around this with minimal changes. 

Alternatively, I have come across the _explicit routing_ of the bottle library: (see [Request Routing -> EXPLICIT ROUTING CONFIGURATION](https://bottlepy.org/docs/dev/routing.html#explicit-routing-configuration)). This seems to allow defining the routes within a function, which could technically solve the import order issue. 

```python
# Taken from the bottle docs:
#  see https://bottlepy.org/docs/dev/routing.html#explicit-routing-configuration

def setup_routing(app):
    app.route('/new', ['GET', 'POST'], form_new)
    app.route('/edit', ['GET', 'POST'], form_edit)

app = Bottle()
setup_routing(app)
```

The only thing is that I am a bit wary of coupling our code to the bottle's interface even more. And if that is to be the solution, I'd like to pick this up in a separate PR to keep the changes isolated. 

But feel to give it a try, maybe that makes it work with `config`, and I will also try to get myself some time to try it out in code too :) 

Furthermore, not to make many changes to the internals of `pypiserver`, it could be a compromise to use the explicit routing approach, but on the client side, when configuring your `pypiserver` instance. A little similar to what I did with adding hooks after the app creation in a separate use case ([this `get_app` on the client side](https://github.com/dee-me-tree-or-love/pypiserver-appengine/blob/master/main.py#L399)). This could also be a way to go around this, making all the changes purely on the client side :) 

In any case, I'll get back to this soon, thanks a lot for all your work! 👍

**Comment by Mason-Lin** [comment]
Hi @dee-me-tree-or-love ,

Thank you for the suggestions.
I'll find time next week to look into these solutions. 😀

**Comment by Mason-Lin** [comment]
Thanks again for the suggestion
Using _explicit routing_ can solve the problem I had before 👍

The practice of `get_app()` mentioned to set `entrypoint: gunicorn -b :8081 -w 2 'main:get_app()'`
I haven't experimented with this method yet, but it seems that adding a health endpoint can be done without modifying `pypiserver`.
Once get the `app` from pypiserver
The user should be able to patch their endpoint from the client side.
Making all the changes purely on the client side.

My current usage is actually to deploy `pypiserver/pypiserver:latest` with docker
Then start a private PyPI server in the private network through CMD ["run", "--disable-fallback"]
I wish I could keep using `entrypoint.sh` and passing some more parameters to accomplish the requirement.
I hope the customized health endpoint becomes a feature out of the box.

I will now change the environment variable approach to using the CLI to set the health endpoint. 😊

**Comment by Mason-Lin** [comment]
notes:
build fail related to #439

**Comment by dee-me-tree-or-love** [comment]
Hey, @Mason-Lin I've fixed the #439 issue with #444. The CI should be back to normal if you update from `master`. I will get to reviewing this more properly soon, thanks for all the updates! :)

**Comment by dee-me-tree-or-love** [comment]
Hey @Mason-Lin, thanks a lot for all the updates on this branch! :v: I've had a few things that kept me busy last week, but I'll try to walk through your PR in detail this week :D

**Comment by Mason-Lin** [comment]
> Hey @Mason-Lin, thanks a lot for all the updates on this branch! :v: I've had a few things that kept me busy last week, but I'll try to walk through your PR in detail this week :D 

No problem, thank you 😁

File: pypiserver/config.py
Comment (dee-me-tree-or-love) [inline pypiserver/config.py:400]:
```suggestion
            "Configure a custom liveness endpoint. It always returns 200 OK if the service is up." 
            "Otherwise it means that the service is unhealthy or dead."
```

I think this makes the purpose of this setting a little more clear :)

File: tests/test_app.py
Comment (dee-me-tree-or-love) [inline tests/test_app.py:205]: Hey, these are really nice tests, thanks a lot for adding! 🤩

File: tests/test_app.py
Comment (dee-me-tree-or-love) [inline tests/test_app.py:202]:
```suggestion
    _app = app(root=root.strpath, health_endpoint="/healthz")
    testapp = webtest.TestApp(_app)
```

Just to keep the lines similar to other test setups :)

File: tests/test_app.py
Comment (dee-me-tree-or-love) [inline tests/test_app.py:211]: Also linked to some comments above, I'd say that here it should raise a `RuntimeError` since this route is already taken :)

File: tests/test_config.py
Comment (dee-me-tree-or-love) [inline tests/test_config.py:392]: This case should raise an error, similar to the `case=f"Invalid hash algo: {val}"`, as this would be a more straightforward outcome imho :) Linked to a comment above on `health_endpoint_arg`. And for inspiration on how to set the error cases up, see `_CONFIG_ERROR_CASES`.

File: pypiserver/__init__.py
Comment (dee-me-tree-or-love) [inline pypiserver/__init__.py:8]: I think we should keep the `DEFAULTS` as much as possible to the `config.py` module, so as to prevent the mixture of responsibilities :) I elaborate on it in other comments.

```suggestion
from pypiserver.config import Config, RunConfig, strtobool
```

**Review by dee-me-tree-or-love — CHANGES_REQUESTED** [review]
Hey @Mason-Lin, thanks a lot for your work and adjusting it to the explicit routing setup! 😊 This is already looking nicer, and I think this addition could be an interesting enabler for some extra configurations for the `pypiserver` in the future. I added a few comments with requests to change below. Those are some alignment details to keep this new behaviour as close as possible to other features of `pypiserver`, plus to make sure it's modular enough for future extensions or changes. I think it's getting there! Could you take a look at the comments? :) 

Special thanks for making all the test cases, these are useful and helpful to see! Then, when you finish this up, don't forget to add a note about this in the `Recipes` section of the `README` and update the `help` command output in the `Quickstart` section, this will be nice ;) 👍 🚀

File: pypiserver/config.py
Comment (dee-me-tree-or-love) [reply pypiserver/config.py:400]: The line length in this suggestion might need to be double checked 😁

**Comment by Mason-Lin** [comment]
> Hey @Mason-Lin, thanks a lot for your work and adjusting it to the explicit routing setup! 😊 This is already looking nicer, and I think this addition could be an interesting enabler for some extra configurations for the `pypiserver` in the future. I added a few comments with requests to change below. Those are some alignment details to keep this new behaviour as close as possible to other features of `pypiserver`, plus to make sure it's modular enough for future extensions or changes. I think it's getting there! Could you take a look at the comments? :) 
> 
> 
> 
> Special thanks for making all the test cases, these are useful and helpful to see! Then, when you finish this up, don't forget to add a note about this in the `Recipes` section of the `README` and update the `help` command output in the `Quickstart` section, this will be nice ;) 👍 🚀  

Hi @dee-me-tree-or-love,

Ok no problem, thank you very much for your review.
I have a work in progress that will probably be done this week, and then I'll start working on this PR 😁

File: tests/test_app.py
Comment (Mason-Lin) [reply tests/test_app.py]: I think you have a point, Explicit is better than implicit.
https://github.com/pypiserver/pypiserver/pull/442/commits/55fa396999c07be76ed58cfe4b14d860809368a1

File: pypiserver/__init__.py
Comment (Mason-Lin) [reply pypiserver/__init__.py]: That's a suggestion👍

When route conflicts it will [raise RuntimeError](https://github.com/pypiserver/pypiserver/pull/442/commits/55fa396999c07be76ed58cfe4b14d860809368a1), so the [DEFAULTS is no longer needed](https://github.com/pypiserver/pypiserver/pull/442/commits/d2849d8a48f499628d80d7e015bf6e119cbe7220) during [setup_routes_from_config()](https://github.com/pypiserver/pypiserver/pull/442/commits/f1ea6707739e4720c93430df114926e11c03a731)

File: tests/test_config.py
Comment (Mason-Lin) [reply tests/test_config.py:392]: [raise SystemExit when using invalid args](https://github.com/pypiserver/pypiserver/pull/442/commits/edcf5778e732440f506b1ec1222fe7d78d0fc1c1)

File: pypiserver/config.py
Comment (Mason-Lin) [reply pypiserver/config.py]: [Config: raise SystemExit when using invalid args](https://github.com/pypiserver/pypiserver/pull/442/commits/35df0cfb8fdbd6ec1733788d503dd9117ba90ccd)

**Comment by Mason-Lin** [comment]
Hi @dee-me-tree-or-love ,
Thank you for your good advice and clear explanation of ideas considering project maintenance and user experience :)

**Comment by Mason-Lin** [comment]
Hi @dee-me-tree-or-love ,
Could you enabled the GitHub actions to run on this PR again? so I can fix them before next review. 🫡

File: README.rst
Comment (dee-me-tree-or-love) [inline README.rst:924]:
```suggestion
``pypiserver`` provides a default health endpoint at ``/health``. It always returns
``200 Ok`` if the service is up. Otherwise, it means that the service is not responsive.
```

Some small wording suggestions :)

File: README.rst
Comment (dee-me-tree-or-love) [inline README.rst:928]:
```suggestion
In addition, ``pypiserver`` allows users to customize the health endpoint.
Alphanumeric characters, hyphens, forward slashes and underscores are allowed
and the endpoint should not overlap with any existing routes.
Valid examples: ``/healthz``, ``/health/live-1``, ``/api_health``, ``/action/health``
```

Some more small suggestions :)

File: README.rst
Comment (dee-me-tree-or-love) [inline README.rst:947]:
```suggestion
Try ``curl http://localhost:8080/action/health``
```

Just for consistency, otherwise, RST doesn't render this as code blocks :)

File: tests/test_config.py
Comment (dee-me-tree-or-love) [inline tests/test_config.py:751]: 🤩  great catch! Thanks 👍

File: pypiserver/config.py
Comment (dee-me-tree-or-love) [inline pypiserver/config.py:405]:
```suggestion
            "the service is up. Otherwise, it means that the service is not responsive."
```

Just as an alignment :)

File: tests/test_main.py
Comment (dee-me-tree-or-love) [inline tests/test_main.py:313]: Awesome! Thanks a lot for setting up all these tests!

**Review by dee-me-tree-or-love — DISMISSED** [review]
Hey @Mason-Lin, this is great work! Thanks a lot for the effort, it was a pleasure to discuss this change 👍 Sorry for my late response, last few weeks I had to manage some challenges on the personal side and was largely AFK. But now getting back to this :) I left a few final nitty comments to polish up the readme and the rest, the code looks good to go IMHO 👍

File: pypiserver/__init__.py
Comment (dee-me-tree-or-love) [inline pypiserver/__init__.py:146]: I see the `black` checks are failing here asking for an extra line. 
image

You can run the autoformatting on all files using `black .` I think.

File: tests/test_main.py
Comment (Mason-Lin) [reply tests/test_main.py:313]: I'm a fan of TDD 😄

File: pypiserver/__init__.py
Comment (Mason-Lin) [reply pypiserver/__init__.py:147]: Thank you, I'm sorry for making mistakes with formatting and typing over and over again 😅
I'm a little unclear on what tests should be done on the local side and how.
I later tried to build `venv`, pip installed `dev.pip`, and executed the `check` related command in `ci.yml` step by step
I hope it won't build fail this time

**Comment by Mason-Lin** [comment]
Thanks for the correction in reStructuredText.
I usually use Markdown, I am unfamiliar with reStructuredText, and I misuse syntax ^^"

**Comment by Mason-Lin** [comment]
> Hey @Mason-Lin, this is great work! Thanks a lot for the effort, it was a pleasure to discuss this change 👍 Sorry for my late response, last few weeks I had to manage some challenges on the personal side and was largely AFK. But now getting back to this :) I left a few final nitty comments to polish up the readme and the rest, the code looks good to go IMHO 👍

no problem, thanks for your time 👍

**Comment by Mason-Lin** [comment]
Hi @dee-me-tree-or-love ,
  Do I need to update the branch or keep what it is?
  If it needs to be updated, should I rebase it or merge it?

Thank you.

File: pypiserver/__init__.py
Comment (dee-me-tree-or-love) [reply pypiserver/__init__.py:147]: Hey, no worries! I'm happy it helped! :)) yes, that's a good point actually, thanks for pointing this out, I think it'll be a nice improvement to contributor docs in general (#368) 👍 Now all looks green, great job 😃

File: README.rst
Comment (dee-me-tree-or-love) [inline README.rst:140]:
```suggestion
        Configure a custom liveness endpoint. It always returns 200 Ok if  
        the service is up. Otherwise, it means that the service is not responsive.
```

Just this tiny detail for consistency, otherwise it's great to go! :) Sorry for taking such a long time for the review 😅

**Review by dee-me-tree-or-love — DISMISSED** [review]
Hey @Mason-Lin! Great job, I think this is ready to go, sorry for taking such a long time for this process, but I'll be happy to merge this :) There is one last tiny remark in the code, and regarding the branch update, if you could, that'd indeed be perfect 🙏  

> Hi @dee-me-tree-or-love ,
> Do I need to update the branch or keep what it is?
> If it needs to be updated, should I rebase it or merge it?
>
> Thank you.

You can do either a rebase or merge, there's no strict process on this since we're squashing PRs into the main branch in the end, it will all appear as a single commit borrowing the PR title :)

**Comment by Mason-Lin** [comment]
Don't worry, I've also learned some good ideas along the way, thanks for every suggestion you give.

Regarding the workflow part, because “First-time contributors need a maintainer to approve running workflows”, I have an idea but I don't know if it is suitable or not, that is to provide a dockerfile, which contains mypy, pylint, black, etc.

If I want to confirm whether the PR will be pass the workflow, I can check with docker build.

**Review by dee-me-tree-or-love — APPROVED** [review]
Great work! 👍 :) Merging

**Comment by dee-me-tree-or-love** [comment]
> Regarding the workflow part, because “First-time contributors need a maintainer to approve running workflows”, I have an idea but I don't know if it is suitable or not, that is to provide a dockerfile, which contains mypy, pylint, black, etc.
> 
> If I want to confirm whether the PR will be pass the workflow, I can check with docker build.

Hey @Mason-Lin, I like this idea! I totally agree that some improved flow for checking will be nice :) Maybe a tool like [`pre-commit`](https://pre-commit.com/), [`drkns`](https://github.com/frantzmiccoli/drkns), or even a collection of scripts in a `Makefile` may be less resource-heavy but allow the same as Dockerfile for checking the code styles :) However the Dockerfile will ensure that the environment is consistent and of course, in that sense, it would also be very beneficial 👍 Maybe a combination of these options will be a nice middle ground, thanks for the suggestion, I really like the proposal.

**Comment by Mason-Lin** [comment]
> Hey @Mason-Lin, I like this idea! I totally agree that some improved flow for checking will be nice :) Maybe a tool like [`pre-commit`](https://pre-commit.com/), [`drkns`](https://github.com/frantzmiccoli/drkns), or even a collection of scripts in a `Makefile` may be less resource-heavy but allow the same as Dockerfile for checking the code styles :) However the Dockerfile will ensure that the environment is consistent and of course, in that sense, it would also be very beneficial 👍 Maybe a combination of these options will be a nice middle ground, thanks a lot for the suggestion, I really like this proposal.  

I think your suggestion is good, and after rethinking it, I guess pre-commit is the best fit because it's the easiest to set up.  (For an experienced developer like you, these may not be a problem😅, my thoughts are just for your reference)

The reason is that people who use the pypiserver project do not necessarily have experience using docker and makefile because the most basic usage of pypiserver is to use pip install directly. (from QuickStart)

When developers find a requirement, they may be able to contribute to it, and the project's contribute guide can guide them to complete the PR smoothly with the lowest learning cost.
They don't need to understand [makefile syntax](https://stackoverflow.com/questions/28712585/when-to-use-space-or-tab-in-makefile) or docker or even wsl (if their OS is windows).  Sparks won't be doused by frustration with these pre-set-up processes.

**Pull request merged**
